runtime/gc_glue_java/HeapRegionDescriptorStandardExtension.hpp

```diff
@@ -69,25 +69,71 @@ class MM_HeapRegionDescriptorStandardExtension : public MM_BaseNonVirtual
 	bool
 	initialize(MM_EnvironmentBase *env)
 	{
-		if (NULL == (_unfinalizedObjectLists = MM_UnfinalizedObjectList::newInstanceArray(env, _maxListIndex))) {
+		if (NULL == (_unfinalizedObjectLists = MM_UnfinalizedObjectList::newInstanceArray(env, _maxListIndex, NULL, 0))) {
 			return false;
 		}
 
-		if (NULL == (_ownableSynchronizerObjectLists = MM_OwnableSynchronizerObjectList::newInstanceArray(env, _maxListIndex))) {
+		if (NULL == (_ownableSynchronizerObjectLists = MM_OwnableSynchronizerObjectList::newInstanceArray(env, _maxListIndex, NULL, 0))) {
 			return false;
 		}
 
-		if (NULL == (_continuationObjectLists = MM_ContinuationObjectList::newInstanceArray(env, _maxListIndex))) {
+		if (NULL == (_continuationObjectLists = MM_ContinuationObjectList::newInstanceArray(env, _maxListIndex, NULL, 0))) {
 			return false;
 		}
 
-		if (NULL == (_referenceObjectLists = MM_ReferenceObjectList::newInstanceArray(env, _maxListIndex))) {
+		if (NULL == (_referenceObjectLists = MM_ReferenceObjectList::newInstanceArray(env, _maxListIndex, NULL, 0))) {
 			return false;
 		}
 
 		return true;
 	}
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	bool
+	reinitializeForRestore(MM_EnvironmentBase* env)
+	{
+		MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(env);
+
+		uintptr_t newListCount = extensions->gcThreadCount;
+
+		Assert_MM_true(_maxListIndex > 0);
+
+		if (newListCount > _maxListIndex) {
+			MM_UnfinalizedObjectList *newUnfinalizedObjectLists;
+			MM_OwnableSynchronizerObjectList *newOwnableSynchronizerObjectLists;
+			MM_ContinuationObjectList *newContinuationObjectLists;
+			MM_ReferenceObjectList *newReferenceObjectLists;
+
+			if (NULL == (newUnfinalizedObjectLists = MM_UnfinalizedObjectList::newInstanceArray(env, newListCount, _unfinalizedObjectLists, _maxListIndex))) {
+				return false;
+			}
+
+			if (NULL == (newOwnableSynchronizerObjectLists = MM_OwnableSynchronizerObjectList::newInstanceArray(env, newListCount, _ownableSynchronizerObjectLists, _maxListIndex))) {
+				return false;
```

Comment:
while false should be returned, it's slightly more correct to not return right away but proceed (or jump to tearDown) , to do the replacement, and then the caller should call tearDown, if false is returned.

the point is all new memory allocations before a failure (at this point only newUnfinalizedObjectLists), should be released (the way I suggested or some other ways)

Reply:
teardown on fail (for new allocs) could be done within this method. 
an example: https://github.com/eclipse-openj9/openj9/blob/870f661786efef51f8b3719a09901f8d33cddb9a/runtime/gc_base/GCExtensions.cpp#L146

but in that case, we might need not to do the teardown of the original structs nor set new _maxListIndex, though. in theory, restore could try to proceed with existing structs, but it's just simpler not to do it (if an alloc failed now, it will fail like soon on something else.

Reply:
Updated the code with a different approach, it more subtle with how mem is freed in failure case

Reply:
```C++
...
	tearDown(env); // Release existing lists (inited during VM startup)

        // Updated global reference to point to local inits which failed
	_unfinalizedObjectLists = newUnfinalizedObjectLists;
	_ownableSynchronizerObjectLists = newOwnableSynchronizerObjectLists;
	_continuationObjectLists = newContinuationObjectLists;
	_referenceObjectLists = newReferenceObjectLists;
...
// return false, rely on  VM shutdown to call MM_HeapRegionDescriptorStandardExtension kill and call tearDown again on new references
return rc;
```

Reply:
This should work, let me know if I'm missing something

Reply:
still think it's not fully correct: in failure case this method should release everything it allocated, leave existing structs intact, and let caller decide if it wants to proceed (or not) with existing structs 

if this method assumes that caller will not proceed with existing ones, there this approach is ok (both existing and all structs will eventually be released), but in general, callee should not make assumptions about how caller should handle the failure. 

for caller however is reasonable to expected that callee, if not succeeding, at least will not make things worse (which will happen, since one of pointers will be null)

but then, if it's too complicated, i'm ok with this

Reply:
Totally agree, I didn't consider that.

Reply:
I've implemented the `goto` approach that you mentioned above